persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,519 @@
+            case char2 of
+                '(' -> pure '('
+                ')' -> pure ')'
+                '\\' -> pure '\\'
+                '\"' -> pure '\"'
+                '\'' -> pure '\''
+                _ -> unexpected (Tokens $ char2 :| [])
+        _ -> pure char1
+
+equality :: Parser Token
+equality = label "equality expression" $ do
+    L.lexeme spaceConsumer $ do
+        lhs <- some contentChar
+        _ <- char '='
+        rhs <-
+            choice
+                [ quotation'
+                , sqlLiteral
+                , parentheticalInner
+                , some $ contentChar <|> char '(' <|> char ')'
+                ]
+        pure $ Equality (Text.pack lhs) (Text.pack rhs)
+  where
+    parentheticalInner = do
+        str <- parenthetical'
+        pure . init . drop 1 $ str
+
+sqlTypeName :: Parser String
+sqlTypeName =
+    some $
+        choice
+            [ alphaNumChar
+            , char '_'
+            ]
+
+sqlLiteral :: Parser String
+sqlLiteral = label "SQL literal" $ do
+    quote <- L.lexeme spaceConsumer $ char '\'' *> manyTill charLiteral (char '\'')
+    st <- optional $ do
+        colons <- string "::"
+        tn <- sqlTypeName
+        pure $ colons <> tn
+    pure $
+        mconcat
+            [ "'"
+            , quote
+            , "'"
+            , fromMaybe "" st
+            ]
+
+quotation :: Parser Token
+quotation = label "quotation" $ do
+    str <- L.lexeme spaceConsumer quotation'
+    pure . Quotation $ Text.pack str
+
+quotation' :: Parser String
+quotation' = char '"' *> manyTill charLiteral (char '"')
+
+parenthetical :: Parser Token
+parenthetical = label "parenthetical" $ do
+    str <- L.lexeme spaceConsumer parenthetical'
+    pure . Parenthetical . Text.pack . init . drop 1 $ str
+
+parenthetical' :: Parser String
+parenthetical' = do
+    str <- between (char '(') (char ')') q
+    pure $ "(" ++ str ++ ")"
+  where
+    q = mconcat <$> some (c <|> parenthetical')
+    c = (: []) <$> choice [contentChar, nonLineSpaceChar, char '"']
+
+blockKey :: Parser Token
+blockKey = label "block key" $ do
+    fl <- upperChar
+    rl <- many alphaNumChar
+    pure . BlockKey . Text.pack $ fl : rl
+
+ptext :: Parser Token
+ptext = label "plain token" $ do
+    str <- L.lexeme spaceConsumer $ some contentChar
+    pure . PText . Text.pack $ str
+
+docComment :: Parser Token
+docComment = hidden $ do
+    _ <- string "-- |" <* hspace
+    str <- many (contentChar <|> nonLineSpaceChar)
+    pure . DocComment . Text.pack $ str
+
+comment :: Parser Token
+comment = hidden $ do
+    _ <- (string "--" <|> string "#") <* hspace
+    str <- many (contentChar <|> nonLineSpaceChar)
+    pure . Comment . Text.pack $ str
+
+indentLevelBacktrack :: Parser () -> Parser Pos
+indentLevelBacktrack sc = lookAhead $ do
+    _ <- sc
+    L.indentLevel
+
+indented :: Parser () -> Pos -> (Parser a -> Parser b) -> Parser a -> Parser b
+indented sc ref f p =
+    f . try $ do
```

Comment:
encourage folks to use `fmap` or apply functions themselves if they need rather than trivially applying it for them

```suggestion
indented :: Parser () -> Pos -> Parser a -> Parser a
indented sc ref p =
    try $ do
```

Reply:
:+1: and I can get rid of `indented'` entirely by doing this